src/charm.py

```diff
@@ -530,6 +550,91 @@ def _on_new_vault_kv_client_attached(self, event: NewVaultKvClientAttachedEvent)
         stale_nonces = set(credential_nonces) - set(nonces)
         self.vault_kv.remove_unit_credentials(relation, stale_nonces)
 
+    def _on_certificates_pki_request_relation_joined(self, event: RelationJoinedEvent):
+        """Initializes the PKI backend."""
+        if not self.unit.is_leader():
+            logger.debug("Only leader unit should configure the vault PKI, skipping")
+            return
+        if not self._is_peer_relation_created():
+            logger.debug("Peer relation not created, deferring event")
+            event.defer()
+            return
+        common_name = self._get_config_common_name()
+        if not common_name:
+            logger.error("Common name is not set correctly in the charm configuration")
+            return
+        try:
+            root_token, _ = self._get_initialization_secret_from_peer_relation()
+        except PeerSecretError:
+            logger.debug("Vault initialization secret not set in peer relation, deferring event")
+            event.defer()
+            return
+        vault = Vault(url=self._api_address, ca_cert_path=self._get_ca_cert_location_in_charm())
+        vault.set_token(token=root_token)
+        if not vault.is_api_available():
+            logger.debug("Vault is not available, deferring event")
+            event.defer()
+            return
+        if vault.is_secret_engine_enabled(path=PKI_MOUNT):
+            logger.debug("PKI backend already enabled, skipping")
+            return
+        vault.enable_pki_engine(path=PKI_MOUNT)
+        csr = vault.configure_pki_intermediate_ca(mount=PKI_MOUNT, common_name=common_name)
+        self._pki_requires.request_certificate_creation(
+            certificate_signing_request=csr.encode(), is_ca=True
+        )
+
+    def _on_certificate_pki_request_certificate_available(self, event: CertificateAvailableEvent):
```

Comment:
What happens if "common_name" is changed after relation_joined ?  Do we need a change in relation in that situation ?

Reply:
Agreed, fixed